### PR TITLE
Add -verror-supplements, to let the user define how many supplemental…

### PR DESCRIPTION
**Pull request by JohanEngelen**
… error messages he wants to see.

In particular this prevents having to rebuild with "-v" flag for ... (18 instantiations, -v to show) to show, and without all the other -v noise.

### CODE REVIEW

**Comment by ibuclaw** [comment]
Can you check what other compilers do?

I have my suspicion that gcc's `-fmax-errors=n` already limits the number of supplemental messages.

**Comment by JohanEngelen** [comment]
> Can you check what other compilers do?
> 
> I have my suspicion that gcc's `-fmax-errors=n` already limits the number of supplemental messages.

LDC does the same as DMD.

File: compiler/src/dmd/globals.d [hunk 1]
```diff
@@ -182,6 +182,7 @@ extern (C++) struct Param
     CHECKACTION checkAction = CHECKACTION.D; // action to take when bounds, asserts or switch defaults are violated
 
     uint errorLimit = 20;
+    uint errorSupplementLimit = 6;      // Limit the number of supplemental messages for each error (0 means unlimited)
```
Comment (ljmf00) [inline compiler/src/dmd/globals.d:185]: Why not maintaining the old hard coded limit, 5, instead of 6?

**Review by ljmf00 — APPROVED** [review]
Looks good to me otherwise. Requires squashing commits.

File: compiler/src/dmd/globals.d (hunk 1 above)
Comment (JohanEngelen) [reply compiler/src/dmd/globals.d:185]: It replaces three hardcoded numbers .  5, 6, and 6.  6 won.  ;)

Comment (ljmf00) [reply compiler/src/dmd/globals.d:185]: Ah, fair, didn't catch the other 6 :)

**Comment by ibuclaw** [comment]
> GDC: `-fmax-errors` does not help. https://d.godbolt.org/z/q1h8xorcb
> 
> Isn't this all in the frontend code, thus the same for all compilers? Why do you expect different behavior?

They go through different diagnostics engines - i.e: errors.d is reimplemented to wrap around gcc diagnostics - maybe ldc reuses llvm diagnostics too if the framework has any.

I couldn't remember whether errorSupplemental bumps the gcc-internal printed error count.  Apparently it doesn't. :-)

**Comment by RazvanN7** [comment]
This requires approval from @WalterBright as it adds a new CLI switch.

**Comment by RazvanN7** [comment]
@WalterBright any thoughts on this?

**Comment by JohanEngelen** [comment]
Ping, waiting for comments before putting more work in it.
It's already part of Weka's compiler, so I have little interest in pushing this forward.

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
This needs a rebase, @WalterBright has had plenty of time to responds. Let's get this in.

**Comment by JohanEngelen** [comment]
Rebased, fixed conflicts, please merge.

**Comment by thewilsonator** [comment]
Thanks, you need to update `frontend.h`.

**Comment by JohanEngelen** [comment]
> Thanks, you need to update `frontend.h`.

ah yes, I got a merge conflict, looking like some things were removed such that frontend.h change was no longer necessary, so being confused I removed it. I see it now in the CI result.

**Comment by JohanEngelen** [comment]
> Thanks, you need to update `frontend.h`.

should be fixed now

File: compiler/test/fail_compilation/fail15616a.d [hunk 2]
```diff
@@ -7,10 +7,11 @@ fail_compilation/fail15616a.d(17):                        `fail15616a.foo(int a,
 fail_compilation/fail15616a.d(26):                        `fail15616a.foo(int a, int b, int c)`
 fail_compilation/fail15616a.d(29):                        `fail15616a.foo(string a)`
 fail_compilation/fail15616a.d(32):                        `fail15616a.foo(string a, string b)`
-fail_compilation/fail15616a.d(41):        ... (3 more, -v to show) ...
+fail_compilation\fail15616a.d(35):                        `fail15616a.foo(string a, string b, string c)`
+fail_compilation/fail15616a.d(41):        ... (2 more, -v to show) ...
```
Comment (AndrejMitrovic) [inline compiler/test/fail_compilation/fail15616a.d:11]: For discovery shouldn't this message be updated to say `-verror-supplements=0 to show`?

**Pull request merged**